wallet/dapps-communication.md

```diff
@@ -9,3 +9,256 @@ Initial meta issue: <https://github.com/massalabs/massa-standards/issues/13>
+```
+
+### Constraints to solve by the provided lib
+
+The library needs to provide a secure and efficient way to communicate between the page script and the content script,
+which allows for multiple requests to be processed in parallel.
+
+> _NOTE:_ 
+> - The communication between the web page script and the content script is contained by the security model retained by browsers. More information [here](https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/Content_scripts#communicating_with_the_web_page).
+> - this is also true for the communication between the content script and the background script. More information [here](https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/Content_scripts#communicating_with_background_scripts) or [here](https://developer.chrome.com/docs/extensions/mv3/messaging/).
+
+
+
+To achieve this, we will use:
+
+- An EventTarget attached to different window keys:
+  - A generic massaWalletProvider for messages from the extension to the web page.
+  - One per extension for messages from the web page to the content script.
+- To allow multiple commands to be executed in parallel, we can set a correlation ID to each outgoing request that is
+propagated in the response.
+
+Here's some example code of massa-wallet-provider library that implements this approach in the webpage script:
+
+```typescript
+
+// global event target to use for all wallet provider
+window.massaWalletProvider = new EventTarget()
+registeredProviders = {}
+
+// Register
+window.massaWalletProvider.addEventListener('register', (payload) => {
+  const extensionEventTarget = new EventTarget();
+  window[`massaWalletProvider-${payload.eventTarget}`] = extensionEventTarget;
+  registeredProviders[payload.providerName] = payload.eventTarget;
+})
+
+interface Message {
+  command: string;
+  params: object;
+  requestId: string;
+}
+
+const pendingRequests = new Map<string, ?>();
+
+//send a message from the webpage script to the content script
+function sendMessageToContentScript(provider, command, params, responseCallback) {
+  const requestId = uuidv4();
+  const message: Message = { command, params, requestId };
+  pendingRequests.set(requestId, responseCallback);
+
+  window[`massaWalletProvider-${registeredProviders[provider]}`].dispatchEvent(new CustomEvent('message', { detail: message }));
+}
+
+//receive a response from the content script
+function handleResponseFromContentScript(event) {
+  const { result, error, requestId } = event.detail;
+  const responseCallback = pendingRequests.get(requestId);
+
+  if (responseCallback) {
+    if (error) {
+      responseCallback(new Error(error));
+    } else {
+      responseCallback(null, result);
+    }
+    pendingRequests.delete(requestId);
+  }
+}
+
+window.massaWalletProvider.addEventListener('message', handleResponseFromContentScript);
+```
+
+Because it was requested, here a potential implementation of the `registerAsMassaWalletProvider` in the content script:
```

Comment:
I still don't understand how the callback given to the sign function will be executed.

Reply:
pseudo code implementation in commit fe31ed329ea6400e25cae9d3e687921fe7f996d5